File: ltrs/beta_generator.py
import numpy as np

def generate_beta_set(length, diameter, total_mass, n_classes=6):
    """
    Generate realistic ballistic coefficient (beta) set.

    Parameters:
        length (m): missile length
        diameter (m): missile diameter
        total_mass (kg): total mass of missile
        n_classes (int): number of debris classes

    Returns:
        beta_set (np.array): beta values (kg/m^2)
    """

    # ---- Assumptions ----
    Cd_mean = 1.4  # average drag coefficient (tumbling debris)

    # Reference cross-sectional area (max frontal area)
    A_ref = np.pi * (diameter / 2)**2

    # ---- Step 1: Mass distribution (log scale) ----
    # From very small fragments to large chunks
    m_min = total_mass * 0.0001   # 0.01%
    m_max = total_mass * 0.3      # 30%

    masses = np.logspace(np.log10(m_min), np.log10(m_max), n_classes)

    # ---- Step 2: Area scaling ----
    # Small fragments: larger area-to-mass ratio
    # Large fragments: closer to body cross-section
    area_scale = np.linspace(3.0, 0.5, n_classes)  # heuristic

    areas = A_ref * area_scale

    # ---- Step 3: Compute beta ----
    beta = masses / (Cd_mean * areas)

    # ---- Step 4: Optional Improvements ----
    # Add randomness
    beta *= np.exp(np.random.normal(0, 0.3, size=n_classes))

    return beta, masses, areas

beta, masses, areas = generate_beta_set(
    length=5.0,      # meters
    diameter=0.3,    # meters
    total_mass=300,  # kg
    n_classes=6
)

print("beta:", beta)
print("masses:", masses)
print("areas:", areas)

File: ltrs/beta_generator.cpp
#include <iostream>
#include <vector>
#include <cmath>

std::vector<double> generate_beta_set_random(
    double length,
    double diameter,
    double total_mass,
    int n_classes = 6
) {
    auto beta = generate_beta_set(length, diameter, total_mass, n_classes);

    // Random generator
    std::random_device rd;
    std::mt19937 gen(rd());
    std::normal_distribution<> dist(0.0, 0.3);  // log-normal variation

    for (auto& b : beta) {
        double factor = std::exp(dist(gen));
        b *= factor;
    }

    return beta;
}

std::vector<double> generate_beta_set(
    double length,       // [m]
    double diameter,     // [m]
    double total_mass,   // [kg]
    int n_classes = 6
) {
    const double Cd_mean = 1.4;

    // Reference area (circular cross-section)
    double A_ref = M_PI * std::pow(diameter / 2.0, 2);

    // Mass range (log scale)
    double m_min = total_mass * 0.0001;  // 0.01%
    double m_max = total_mass * 0.3;     // 30%

    std::vector<double> beta_set;
    beta_set.reserve(n_classes);

    for (int i = 0; i < n_classes; i++) {
        // ---- Log-space interpolation ----
        double t = static_cast<double>(i) / (n_classes - 1);

        double mass = std::pow(10.0,
            std::log10(m_min) + t * (std::log10(m_max) - std::log10(m_min))
        );

        // ---- Area scaling (heuristic) ----
        double area_scale = 3.0 - t * (3.0 - 0.5);  // from 3.0 → 0.5
        double area = A_ref * area_scale;

        // ---- Beta calculation ----
        double beta = mass / (Cd_mean * area);

        beta_set.push_back(beta);
    }

    return beta_set;
}

int main() {
    double length = 5.0;     // m
    double diameter = 0.3;   // m
    double total_mass = 300; // kg

    auto beta_set = generate_beta_set(length, diameter, total_mass);

    std::cout << "Beta values:\n";
    for (double b : beta_set) {
        std::cout << b << " kg/m^2\n";
    }

    return 0;
}
